### PR TITLE
IRMutator->IRMutator2, Part 1

### PR DESCRIPTION
**Pull request by steven-johnson**
Convert add_custom_lower_pass() and everything that uses it

### CODE REVIEW

File: test/correctness/gpu_thread_barrier.cpp [hunk 1]
```diff
@@ -21,13 +21,13 @@ class CountBarriers : public IRVisitor {
...
-class CheckBarrierCount : public IRMutator {
+class CheckBarrierCount : public IRMutator2 {
     int correct;
 public:
     CheckBarrierCount(int correct) : correct(correct) {}
-    using IRMutator::mutate;
+    using IRMutator2::mutate;
 
-    Stmt mutate(Stmt s) {
+    Stmt mutate(const Stmt &s) override {
```
Comment (abadams) [inline test/correctness/gpu_thread_barrier.cpp]: Interesting. I think the old version of this was buggy

**Comment by abadams** [comment]
This one is potentially a breaking change for users. We should preserve the ability to use IRMutator too, and mark it as deprecated

File: test/correctness/gpu_thread_barrier.cpp (hunk 1 above)
Comment (steven-johnson) [reply test/correctness/gpu_thread_barrier.cpp]: Yep.

**Comment by steven-johnson** [comment]
> We should preserve the ability to use IRMutator too, and mark it as deprecated

It's pretty trivial to update to the new interface; it would be a lot simpler to just tell people "update your code". That said, I'll see what I can do.

**Comment by steven-johnson** [comment]
_We should preserve the ability to use IRMutator too_

I'll do that as a followup PR (as a trivial adaptor class, at least).

**Pull request merged**